servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/ClientIpUtils.java

```diff
@@ -0,0 +1,215 @@
+                                        final ConnectionContext ctx,
+                                        final boolean fallbackToSocketAddress) {
+
+        final CharSequence ip = parseXForwardedFor(headers.get(X_FORWARDED_FOR));
+        if (ip != null) {
+            return ip;
+        }
+
+        final CharSequence ipFromForwarded = parseForwarded(headers.get(FORWARDED));
+        if (ipFromForwarded != null) {
+            return ipFromForwarded;
+        }
+
+        return fallbackToSocketAddress ? toHostString(ctx.remoteAddress()) : null;
+    }
+
+    /**
+     * Retrieves the first IP in a comma+space separated list of IPs, or the value as-is if it contains no comma.
+     * <p>
+     * This is useful for cases when the IP is retrieved from the {@link HttpHeaderNames#X_FORWARDED_FOR} which value is
+     * of the form: {@code client, proxy1, proxy2, ...}
+     *
+     * @param value a value of the header, which could contain client IP
+     * @return extracted client IP address or {@code null}
+     */
+    @Nullable
+    private static CharSequence parseXForwardedFor(@Nullable final CharSequence value) {
+        if (value == null || value.length() == 0) {
+            return null;
+        }
+
+        final int idx = indexOf(value, ',');
+        final CharSequence ip = idx < 0 ? value : value.subSequence(0, idx);
+        return ip.length() == 0 ? null : ip;
+    }
+
+    /**
+     * Parses a value of {@link HttpHeaderNames#FORWARDED} header to extract legal IP address. For more information, see
+     * <a href="https://tools.ietf.org/html/rfc7239">RFC 7239: Forwarded HTTP Extension</a>.
+     *
+     * @param value a value of {@link HttpHeaderNames#FORWARDED} header
+     * @return extracted client IP address or {@code null} if not presented
+     */
+    @Nullable
+    private static CharSequence parseForwarded(@Nullable final CharSequence value) {
+        if (value == null || value.length() < 11) {
+            // 11 is the shortest length of "Forwarded" header, which may contain a valid origin IP address:
+            // 'for="[::1]"' or 'for=0.0.0.0'
+            return null;
+        }
+
+        final String str = value.toString().toLowerCase();  // required for case-insensitive indexOf
+        int beginIdx = str.indexOf("for=");
+        if (beginIdx < 0) {
+            return null;
+        }
+
+        beginIdx += 4;
+        if (str.length() - beginIdx < 7) {
+            // not enough data to extract the shortest valid origin IP address: "[::1]" or 0.0.0.0
+            return null;
+        }
+
+        int endIdx = computeForParamEndIdx(str, beginIdx);
+        if (endIdx - beginIdx < 7) {
+            // not enough data to extract the shortest valid origin IP address: "[::1]" or 0.0.0.0
+            return null;
+        }
+
+        if (str.charAt(beginIdx) == '"') {
+            // unwrap quoted value
+            beginIdx++;
+            endIdx--;
```

Comment:
Worth checking if the corresponding character is also `"`. If it's improperly formatted it might escape some prior validation and can be used to trick the util to read a spoofed IP.